LOC: request device configuration and publish metrics/info

rouming opened the pull request:
This work introduces two LOC related things if LOC configuration exists (LOC URL is set by a controller):

1. Device configuration can be fetched from another source, aka LOC, if main controller is unreachable or returns an error.

3. /info and /metrics are published to the LOC URL along with publishing to a controller. The main difference is that /info is made recurring for the LOC case, because LOC is supposed to be stateless so all /info structures are published without any sending error handling.

Most of the commits are API changes to support different destinations for publishing requests: controller, loc, local server.

eriknordmark commented:
> * [cca1885](https://github.com/lf-edge/eve/commit/cca188572372d6c62cb94fd36c5d4c58059ff09d) handleconfig: handle 'CertInvalid' and 'CertMiss' similar in the getLatestConfig()
>   Do we really handle 'CertMiss' and 'CertInvalid' differently? I unify the logic.

They are different things; CertMiss is about the controller (signing) certificate, and CertInvalid is about the TLS certificate chain. We should probably add that information in the comments for the enums to make this more clear.

eriknordmark commented on pkg/pillar/cmd/client/client.go:
```diff
@@ -494,23 +494,10 @@ func myPost(zedcloudCtx *zedcloud.ZedCloudContext, tlsConfig *tls.Config,
-			}
-		default:
-			log.Error(err)
+		log.Errorf("%s, err %v", rv.Status.String(), err)
```
We intentionally had different log levels here because we don't need to make noise around a controller upgrade causing it to return Upgrade or Refused, while we do want to know about CertInvalid.

rouming replied:
Commit is removed.

eriknordmark commented on pkg/pillar/cmd/diag/diag.go:
```diff
@@ -1052,26 +1052,8 @@ func myGet(ctx *diagContext, reqURL string, ifname string,
-			fmt.Fprintf(outfile, "ERROR: %s: get %s failed: %s\n",
-				ifname, reqURL, err)
-		}
+		fmt.Fprintf(outfile, "ERROR: %s, %s: %s, failed: %v",
+			ifname, reqURL, rv.Status.String(), err)
```
Makes sense to use the new String here to simplify.

rouming replied:
Commit is removed.

eriknordmark commented on pkg/pillar/cmd/zedagent/handlecertconfig.go:
```diff
@@ -255,18 +255,7 @@ func getCertsFromController(ctx *zedagentContext, desc string) (success bool) {
-		default:
-			log.Errorf("getCertsFromController failed: %s", err)
-		}
+		log.Errorf("getCertsFromController: %s, failed: %s", rv.Status.String(), err)
```
Should preserve different log levels I think.

rouming replied:
Commit is removed.

eriknordmark commented on pkg/pillar/cmd/zedagent/handleconfig.go:
```diff
@@ -185,12 +186,15 @@ func (s configSource) String() string {
+	invalidAuthContainer                        // config is not signed by the controller
+	obsoleteConfig                              // newer config is already applied
+	malformedConfig                             // unmarshaling failed
+	invalidConfig                               // config is not valid (cannot be parsed, UUID mismatch, bad signature, etc.)
```
signature failure is listed for two of the values. Which one is used?

rouming replied:
Commit is removed.

eriknordmark commented on pkg/pillar/cmd/zedagent/parseconfig.go:
```diff
@@ -2512,6 +2509,17 @@ func parseOpCmds(getconfigCtx *getconfigContext,
+func parseLocConfig(getconfigCtx *getconfigContext,
+	config *zconfig.EdgeDevConfig) {
+	locConfig := config.GetLocConfig()
+	if locConfig != nil && len(locConfig.LocUrl) > 0 {
```
Shouldn't the controller be able to clear getconfigCtx.locConfig by passing nil?

rouming replied:
There is a check for null.

eriknordmark replied:
Yes, but when nil you don't update  getconfigCtx.locConfig to be nil, hence I don't see how the controller can disable it.

eriknordmark commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -2356,17 +2356,8 @@ func getDeferredSentHandlerFunction(ctx *zedagentContext) *zedcloud.SentHandlerF
-					log.Functionf("sendAttestReqProtobuf: Controller certificate miss")
-				case types.SenderStatusNotFound:
-					log.Functionf("sendAttestReqProtobuf: Controller SenderStatusNotFound")
+				log.Functionf("sendAttestReqProtobuf: %s", result.String())
```
Should have different log levels here.

rouming replied:
Commit removed

eriknordmark reviewed (changes requested):
Some detailed comments in the diffs.

rouming commented:
Difference to the previous version:
* Unite two LOC related works in one PR: fetch config, publish /info, /metrics
* Remove config refactoring commits as not very much helpful

rouming commented:
> > * [cca1885](https://github.com/lf-edge/eve/commit/cca188572372d6c62cb94fd36c5d4c58059ff09d) handleconfig: handle 'CertInvalid' and 'CertMiss' similar in the getLatestConfig()
> >   Do we really handle 'CertMiss' and 'CertInvalid' differently? I unify the logic.
> 
> They are different things; CertMiss is about the controller (signing) certificate, and CertInvalid is about the TLS certificate chain. We should probably add that information in the comments for the enums to make this more clear.

Commit is removed.

milan-zededa commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -235,10 +235,68 @@ var logger *logrus.Logger
+// destination types, where info should be sent
+const (
+	ControllerDest  infoDest = 1
+	LocalServerDest          = 2
```
Let's maybe start consistently denoting Local Profile Server with LPS (similalry to LOC abbreviation). Se here I would suggest `LPSDest` instead.

rouming replied:
Yep. Do we have shorter name for the "Controller"?

eriknordmark replied:
No, we don't. But LPS is more accurate that LocalServer.

milan-zededa commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -112,8 +111,9 @@ type zedagentContext struct {
-	TriggerHwInfo             chan<- struct{}
+	TriggerDeviceInfo         chan<- infoDest
+	TriggerHwInfo             chan<- infoDest
+	TriggerCloudLocationInfo  chan<- infoDest
```
All these Trigger* fields can start with lower-case, zedagent context is not exported (I realize this mistake was there already).

rouming replied:
Yep.

milan-zededa commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -235,10 +235,68 @@ var logger *logrus.Logger
 var log *base.LogObject
 var zedcloudCtx *zedcloud.ZedCloudContext
 
+type infoDest uint
```
What about `publishDest` instead to make it more clear and cover metrics as well?

milan-zededa replied:
And maybe also with `s` or `Set` suffix to make it clear that a value of this type is essentially a set, built using bit flags.

rouming replied:
Bitset?

rouming replied:
The suffix, infoDestBitset

eriknordmark replied:
+1 to use a different name to cover metrics as well. publishDest or apiDest are names to consider.

milan-zededa commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -235,10 +235,68 @@ var logger *logrus.Logger
+			withNetTracing, itemType, flags)
+		items = append(items, item)
+	}
+
```
Do you plan to later also implement `LocalServerDest` here?

rouming replied:
Not now. Did not want to touch this part.

eriknordmark replied:
But you already have AllDest include LocalServerDest yet you only process that for the DeviceInfo. So it makes it hard to understand what does AllDest mean since it is type dependent what the code actually does.

milan-zededa commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -1786,42 +1846,57 @@ func triggerPublishHwInfo(ctxPtr *zedagentContext) {
 }
 
-func triggerPublishAllInfo(ctxPtr *zedagentContext) {
+func triggerPublishLocationToController(ctxPtr *zedagentContext) {
```
Shouldn't this be `triggerPublishLocation` (without `ToController`)?
Also `triggerPublishLocationToControllerWithDest` should be probably just `triggerPublishLocationWithDest`

rouming replied:
Yep, can be removed.

milan-zededa commented on pkg/pillar/cmd/zedagent/parseconfig.go:
```diff
@@ -2512,6 +2516,17 @@ func parseOpCmds(getconfigCtx *getconfigContext,
+	locConfig := config.GetLocConfig()
+	if locConfig != nil && len(locConfig.LocUrl) > 0 {
+		getconfigCtx.locConfig = &types.LOCConfig{
+			LocURL: locConfig.LocUrl,
```
Should we try to parse the URL (using `url.Parse()`) for validation?

milan-zededa reviewed (approved):
Starting eden tests, but left some review comments + there are some revive issues worth fixing.

eriknordmark reviewed (approved):
Kick of regression tests  

Are there some eden tests being added for the LOC part?

eriknordmark commented:
Dco is saying some commit is not signed

eriknordmark commented on pkg/pillar/cmd/zedagent/reportinfo.go:
```diff
@@ -45,7 +45,7 @@ var (
 )
 
-func deviceInfoTask(ctxPtr *zedagentContext, triggerDeviceInfo <-chan struct{}) {
+func deviceInfoTask(ctxPtr *zedagentContext, triggerDeviceInfo <-chan infoDest) {
```
I'm consuded by having infoDest be a type in some functions like this one, and a variable name in other functions. Can you disambiguate the names?

eriknordmark commented on pkg/pillar/cmd/zedagent/reportinfo.go:
```diff
@@ -707,9 +714,12 @@ func PublishAppInstMetaDataToZedCloud(ctx *zedagentContext, appInstMetadata *typ
+	ditems := createDeferredItemsByDest(ctx, dest, ignoreErr, deferKey, buf,
+		size, true, false, info.ZInfoTypes_ZiAppInstMetaData)
+	zedcloud.AddDeferredList(zedcloudCtx, ditems)
+	zedcloud.HandleDeferred(zedcloudCtx, time.Now(), 0, false)
```
Why are you changing the last argument to HandleDeferred from true to false here?

eriknordmark commented on pkg/pillar/cmd/zedagent/reportinfo.go:
```diff
@@ -654,14 +658,19 @@ func PublishDeviceInfoToZedCloud(ctx *zedagentContext) {
+	ditems := createDeferredItemsByDest(ctx, dest, ignoreErr, deviceUUID, buf,
+		size, true, withNetTracing, info.ZInfoTypes_ZiDevice)
+	zedcloud.AddDeferredList(zedcloudCtx, ditems)
+	zedcloud.HandleDeferred(zedcloudCtx, time.Now(), 0, false)
```
Why are you changing the last argument to HandleDeferred from true to false here?

eriknordmark commented on pkg/pillar/cmd/zedagent/hardwareinfo.go:
```diff
@@ -98,17 +98,17 @@ func PublishHardwareInfoToZedCloud(ctx *zedagentContext) {
+	ditems := createDeferredItemsByDest(ctx, dest, ignoreErr, hwInfoKey, buf,
+		size, bailOnHTTPErr, false, info.ZInfoTypes_ZiHardware)
+	zedcloud.AddDeferredList(zedcloudCtx, ditems)
+	zedcloud.HandleDeferred(zedcloudCtx, time.Now(), 0, false)
```
Why are you changing the last argument to HandleDeferred from true to false here?

eriknordmark commented on pkg/pillar/cmd/zedagent/handlenetworkinstance.go:
```diff
@@ -180,9 +180,11 @@ func prepareAndPublishNetworkInstanceInfoMsg(ctx *zedagentContext,
 		true, false, zinfo.ZInfoTypes_ZiNetworkInstance)
-	zedcloud.HandleDeferred(zedcloudCtx, time.Now(), 0, true)
+	zedcloud.AddDeferredList(zedcloudCtx, ditems)
+	zedcloud.HandleDeferred(zedcloudCtx, time.Now(), 0, false)
```
Why are you changing the last argument to HandleDeferred from true to false here?

eriknordmark commented on pkg/pillar/cmd/zedagent/handlemetrics.go:
```diff
@@ -1471,9 +1483,11 @@ func PublishEdgeviewToZedCloud(ctx *zedagentContext, evStatus *types.EdgeviewSta
 		true, false, info.ZInfoTypes_ZiEdgeview)
-	zedcloud.HandleDeferred(zedcloudCtx, time.Now(), 0, true)
+	zedcloud.AddDeferredList(zedcloudCtx, ditems)
+	zedcloud.HandleDeferred(zedcloudCtx, time.Now(), 0, false)
```
Why are you changing the last argument to HandleDeferred from true to false here?

eriknordmark commented on pkg/pillar/cmd/zedagent/handlemetrics.go:
```diff
@@ -1421,13 +1431,16 @@ func PublishBlobInfoToZedCloud(ctx *zedagentContext, blobSha string, blobStatus
 		true, false, info.ZInfoTypes_ZiBlobList)
-	zedcloud.HandleDeferred(zedcloudCtx, time.Now(), 0, true)
+	zedcloud.AddDeferredList(zedcloudCtx, ditems)
+	zedcloud.HandleDeferred(zedcloudCtx, time.Now(), 0, false)
```
Why are you changing the last argument to HandleDeferred from true to false here?

eriknordmark commented on pkg/pillar/cmd/zedagent/handlemetrics.go:
```diff
@@ -1362,15 +1370,18 @@ func PublishVolumeToZedCloud(ctx *zedagentContext, uuid string,
 		true, false, info.ZInfoTypes_ZiVolume)
-	zedcloud.HandleDeferred(zedcloudCtx, time.Now(), 0, true)
+	zedcloud.AddDeferredList(zedcloudCtx, ditems)
+	zedcloud.HandleDeferred(zedcloudCtx, time.Now(), 0, false)
```
Why are you changing the last argument to HandleDeferred from true to false here?

eriknordmark commented on pkg/pillar/cmd/zedagent/handlemetrics.go:
```diff
@@ -1283,16 +1290,18 @@ func PublishContentInfoToZedCloud(ctx *zedagentContext, uuid string,
 		true, false, info.ZInfoTypes_ZiContentTree)
-	zedcloud.HandleDeferred(zedcloudCtx, time.Now(), 0, true)
+	zedcloud.AddDeferredList(zedcloudCtx, ditems)
+	zedcloud.HandleDeferred(zedcloudCtx, time.Now(), 0, false)
```
Why are you changing the last argument to HandleDeferred from true to false here?

eriknordmark commented on pkg/pillar/cmd/zedagent/handlemetrics.go:
```diff
@@ -1213,16 +1219,18 @@ func PublishAppInfoToZedCloud(ctx *zedagentContext, uuid string,
 		true, false, info.ZInfoTypes_ZiApp)
-	zedcloud.HandleDeferred(zedcloudCtx, time.Now(), 0, true)
+	zedcloud.AddDeferredList(zedcloudCtx, ditems)
+	zedcloud.HandleDeferred(zedcloudCtx, time.Now(), 0, false)
```
Why are you changing the last argument to HandleDeferred from true to false here?

eriknordmark replied:
See above; let's talk about this.

eriknordmark commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -235,10 +235,68 @@ var logger *logrus.Logger
 type infoForObjectKey struct {
 	infoType  info.ZInfoTypes
 	objectKey string
+	infoDest  infoDest
```
field name and type being named the same makes the code harder to understand.

eriknordmark commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -235,10 +235,68 @@ var logger *logrus.Logger
+	return flags
+}
+
+func createDeferredItemsByDest(ctx *zedagentContext, dest infoDest,
```
Would be helpful to have a comment saying what this function does. Returns zero or more deferred items based on the destination.

eriknordmark commented on pkg/pillar/cmd/zedagent/handlelocation.go:
```diff
@@ -59,14 +60,17 @@ func locationTimerTask(ctx *zedagentContext, handleChannel chan interface{}) {
 		select {
 		case <-cloudTicker.C:
+			// Loopback the destination, will be fetched on the next iteration
+			triggerCloudLocationInfo <- AllDest
```
I'm confused by the naming. What does the "cloud" part of CloudLocationInfo supposed to mean when you are sending this to all destinations? Shouldn't it be named triggerLocationInfo?

eriknordmark replied:
How do you know that this can not block?
If the triggerPublishLocationToControllerWithDest() sent one message to the channel and then before that is processed this function handles a ticker, then it will block until the channel is read, but only this function reads the channel.

One way to solve this is to make this a conditional send but not clear to me that will always use the correct set of destinations.

eriknordmark commented on pkg/pillar/cmd/zedagent/handlemetrics.go:
```diff
@@ -164,10 +165,10 @@ func encodeTestResults(tr types.TestResults) *info.ErrorInfo {
 
-// Run a periodic post of the metrics
-func metricsTimerTask(ctx *zedagentContext, handleChannel chan interface{}) {
+// Run a periodic post of the metrics and info
```
Would be more clear to say "and info to an LOC if we have one".

eriknordmark commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -407,13 +467,13 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject, ar
 	// start the metrics reporting task
 	handleChannel := make(chan interface{})
 	log.Functionf("Creating %s at %s", "metricsTimerTask", agentlog.GetMyStack())
```
Missing rename

eriknordmark commented on pkg/pillar/cmd/zedagent/handlemetrics.go:
```diff
@@ -1536,32 +1550,47 @@ func SendProtobuf(url string, buf *bytes.Buffer, size int64,
 	}
 }
 
+func sendMetricsProtobuf(ctx *getconfigContext,
```
One of the implicit assumptions around the stillrunning logic for watchdogs is that each is that we only do one attempt to send before we call stillrunning again. This is because the http/tcp timeout is 2 minutes and the watchdog will trigger at 5 minutes so doing more than one send attempt is getting close to a spurious watchdog.

So I think the check for ctx.locConfig needs to be in the loop in metricsTimerTask.

eriknordmark commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -235,10 +235,68 @@ var logger *logrus.Logger
 
+type infoDest uint
+
+// destination types, where info should be sent
```
And update comment to not be specific to info messages.

eriknordmark commented on pkg/pillar/zedcloud/deferred.go:
```diff
@@ -52,8 +60,12 @@ type DeferredContext struct {
+
 // TypePriorityCheckFunction returns true in case of find type with high priority
-type TypePriorityCheckFunction func(itemType interface{}) bool
+type TypePriorityCheckFunction func(itemType interface{}, flags DeferredItemFlags) bool
```
The flags argument is not used. Is that intentional?

eriknordmark commented on pkg/pillar/zedcloud/deferred.go:
```diff
@@ -223,46 +237,61 @@ func (ctx *DeferredContext) handleDeferred(log *base.LogObject, event time.Time,
-func (ctx *DeferredContext) setDeferred(zedcloudCtx *ZedCloudContext,
-	key string, buf *bytes.Buffer, size int64, url string, bailOnHTTPErr,
-	withNetTracing bool, itemType interface{}) {
+func (ctx *DeferredContext) addDeferred(zedcloudCtx *ZedCloudContext,
```
I didn't check if both AddDeferred and SetDeferred are still used, but they intentionally are different since Set replaces an item with the same key and Add does not. So renaming "set"to "add" while keeping the replace semantics of "set" doesn't make sense to me.

eriknordmark replied:
Turns out that AddDeferred (queue without replacement) is gone except one mention in a comment.
But since your new AddDeferredList has replacement semantics it makes sense naming it SetDeferredList, or carefully change all comments around the replacement semantics to make sure that we state that Add is doing a replace if the key already exists.

eriknordmark commented on pkg/pillar/cmd/zedagent/handlelocation.go:
```diff
@@ -148,15 +151,16 @@ func publishLocationToController(locInfo *info.ZInfoLocation, iteration int) {
+	items := createDeferredItemsByDest(ctx, dest, ignoreErr, key, buf, size,
+		bailOnHTTPErr, withNetTrace, info.ZInfoTypes_ZiLocation)
+	zedcloud.AddDeferredList(zedcloudCtx, items)
+	zedcloud.HandleDeferred(zedcloudCtx, time.Now(), 0, false)
```
I think this should at most send len(items) thus instead of a sendOne bool we need to make HandleDeferred that a sendCount int, which you set to len(items) across the board.
Basically we only try to send as many items as were added by this function.

eriknordmark reviewed (changes requested):
I realize the interaction between the 2 minute http/tcp timeout, the 5 minute stillrunning watchdog, and the SetDeferred+HandleDeferred only processing one message is a bit subtle. So far that has been done by having such a HandleDeferred be called with sendOne=true to make sure it doesn't spend more than 2 minutes total.

If we want to send multiple we either need to convince ourselves that we stll have margins in the worst case so we don't get close to the 5 minute watchdog firing due to lack of stillrunning calls, or we need to pass in the strillrunning to HandleDeferred so it can be called for after each attempt to send.

Let's talk about this next week.

rouming commented:
Differences to the previous version:
* Refactor deferred logic by creating another deferred context for all recurring (periodic) requests. The periodic deferred queue is handled separately from another goroutine in order not to block the main event loop and not to interfere with /attest requests. Also all requests to the periodic deferred queue ignore all errors.
* Changes based on Milan and Erik comments: function renames, type renames, comments tweaks, private structure members, etc

eriknordmark commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -656,9 +714,10 @@ func waitUntilDNSReady(zedagentCtx *zedagentContext, stillRunning *time.Ticker)
-			zedcloud.HandleDeferred(zedcloudCtx, change, 100*time.Millisecond, false)
+			zedcloudCtx.DeferredEventCtx.HandleDeferred(
+				change, 100*time.Millisecond, false)
 			zedagentCtx.ps.CheckMaxTimeTopic(agentName, "deferredChan", start,
```
Want to rename the deferredChan to deferredEventCtx?

rouming replied:
Done.

eriknordmark commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -738,8 +810,10 @@ func mainEventLoop(zedagentCtx *zedagentContext, stillRunning *time.Ticker) {
+				zedcloudCtx.DeferredEventCtx.HandleDeferred(
+					start, 100*time.Millisecond, false)
+				zedagentCtx.ps.CheckMaxTimeTopic(agentName,
+					"deferredChan", start, warningTime, errorTime)
```
Ditto?

rouming replied:
Done.

eriknordmark commented on pkg/pillar/zedcloud/deferred.go:
```diff
@@ -17,15 +17,14 @@ import (
-//      case change := <- deferredChan:
-//		zedcloud.HandleDeferred(zedcloudCtx, change)
+//      case change := <- ctx.Ticker.C:
+//          ctx.HandleDeferred(...)
```
Nit but the change aka event argument is still there in HandleDeferred.

rouming replied:
I put '...' to avoid repeating all the arguments. Seems there is no need in such an example to expose exact arguments we have for this function.

eriknordmark commented on pkg/pillar/zedcloud/deferred.go:
```diff
@@ -137,15 +138,15 @@ func (ctx *DeferredContext) handleDeferred(log *base.LogObject, event time.Time,
 					key, rv.Status, err)
-				exit = true
+				exit = !item.ignoreErr
 				// Make sure we pass a non-zero result to the sentHandler.
 				if rv.Status == types.SenderStatusNone {
 					rv.Status = types.SenderStatusFailed
 				}
 			} else if rv.Status != types.SenderStatusNone {
 				log.Functionf("handleDeferred: for %s received unexpected status %d",
 					key, rv.Status)
-				exit = true
+				exit = !item.ignoreErr
```
When we ignore the error and do not exit, for consistency shouldn't we also avoid updating rv.Status? Or should the definition be that the sentHandler can see the error, but we do not exitOnError? (If so it might make sense to rename or at least document what item.ignoreErr means).

rouming replied:
I do not think this is a good idea to pretend no error happened. We say 'there is an error' but continue processing the queue. The intention is not to halt the queue and ignore the error.

eriknordmark replied:
So the semantics are ignoreButLogError and not completelyIgnoreError. It would make sense to document that e.g., where the ignoreErr field is defined in the struct.

rouming replied:
Will add a comment for the SetDeferred() function where this parameter is passed.

eriknordmark replied:
Did you already add that?

eriknordmark commented on pkg/pillar/cmd/zedagent/handleconfig.go:
```diff
@@ -579,9 +575,17 @@ func getLatestConfig(getconfigCtx *getconfigContext, url string,
 						ts.Format(time.RFC3339Nano))
+
+					cfgRetval := inhaleDeviceConfig(getconfigCtx, config, savedConfig)
+					if cfgRetval != configOK {
+						log.Errorf("inhaleDeviceConfig failed: %d", cfgRetval)
+						return cfgRetval, rv.TracedReqs
+					}
```
Have you seen the inhale fail on a saved config? 
I don't have an issue with this added check, but wonder if there is some unknown issue which causes failures.

eriknordmark replied:
Ah this is due to obsoleteConfig. Makes sense.

eriknordmark commented on pkg/pillar/cmd/zedagent/handleconfig.go:
```diff
@@ -643,27 +647,66 @@ func getLatestConfig(getconfigCtx *getconfigContext, url string,
+	return configOK, rv.TracedReqs
+}
+
+// Returns true if attempt to get a configugration has failed, but initial
+// configuration was received (either from the controller, either successfully
+// read from the file)
+func needRequestLocConfig(getconfigCtx *getconfigContext,
+	rv configProcessingRetval) bool {
+
+	return (rv != configOK && getconfigCtx.locConfig != nil)
+}
```
First I had to read this comment and function carefully to understand its semantics, then go look at where it is called/used, only to realize that it is only called in one place and if the logical expression was inline I would have saved 2 minutes in reading.

This is what I call a too shallow abstraction. I know that is very subjective.

eriknordmark commented on pkg/pillar/cmd/zedagent/handlelocation.go:
```diff
@@ -59,28 +60,11 @@ func locationTimerTask(ctx *zedagentContext, handleChannel chan interface{}) {
-			publishLocationToController(locInfo, cloudIteration)
-			ctx.ps.CheckMaxTimeTopic(wdName, "publishLocationToController", start,
-				warningTime, errorTime)
-
+			publishLocation(ctx, &cloudIteration, wdName, ControllerDest)
```
@milan-zededa don't we want to publish periodically to AllDest here? (and, if so, rename the tcloudTicker variable)?

eriknordmark replied:
With the current code in the PR we only publish location to the LOC when we publish all info which is when the controller epoch changes.

eriknordmark replied:
But see below - I might have missed a new use of triggerPublishAllInfo(). Makes sense to add comments about that new usage here resulting in publishing the location and check that the comment for triggerPublishAllInfo isn't stating something misleading abouts its usage.

rouming replied:
No sure I understand comments right. The default behavior is unchanged. If not - this is a bug.

eriknordmark commented on pkg/pillar/cmd/zedagent/handlemetrics.go:
```diff
@@ -193,6 +194,12 @@ func metricsTimerTask(ctx *zedagentContext, handleChannel chan interface{}) {
 			ctx.ps.CheckMaxTimeTopic(wdName, "publishMetrics", start,
 				warningTime, errorTime)
 
```
Makes sense to call 		ctx.ps.StillRunning(wdName, warningTime, errorTime)
here since we are doing a second send attempt to LOC.

rouming replied:
Done.

eriknordmark commented on pkg/pillar/cmd/zedagent/handlemetrics.go:
```diff
@@ -193,6 +194,12 @@ func metricsTimerTask(ctx *zedagentContext, handleChannel chan interface{}) {
 			ctx.ps.CheckMaxTimeTopic(wdName, "publishMetrics", start,
 				warningTime, errorTime)
 
+			locConfig := ctx.getconfigCtx.locConfig
+			if locConfig != nil {
+				// Publish all info by timer only for LOC
+				triggerPublishAllInfo(ctx, LOCDest)
+			}
```
At some point in time we might want a separate setable timer value for the periodic info to LOC, but we can worry about that later I think.

eriknordmark replied:
Does this triggerPublishAllInfo cause the location to be publishes as well? Then we should document that where I asked about the location publishing above.

milan-zededa replied:
Since we already have separate timers for location publishing, I would prefer it those were respected and location wasn't published alongside info or metrics.

rouming replied:
I did not change the default behavior. If I did - this is a bug. So metrics timer is abused (yes) and used for publishing location info only for the LOC. `locationTimerTask` publishes location from the timer only for the controller and LPS (as it was before).

eriknordmark commented on pkg/pillar/zedcloud/deferred.go:
```diff
@@ -220,20 +221,13 @@ func (ctx *DeferredContext) handleDeferred(log *base.LogObject, event time.Time,
-}
-
-func (ctx *DeferredContext) setDeferred(zedcloudCtx *ZedCloudContext,
+func (ctx *DeferredContext) SetDeferred(
```
Note that there are a few revive comments from yetus visible in the diffs.

eriknordmark commented on pkg/pillar/zedcloud/deferred.go:
```diff
@@ -291,17 +282,22 @@ func (ctx *DeferredContext) removeDeferred(zedcloudCtx *ZedCloudContext, key str
 }
 
+// Kicks the timer for immediate execution
+func (ctx *DeferredContext) KickTimer() {
```
Note that there are a few revive comments from yetus visible in the diffs.

eriknordmark reviewed (approved):
Looks very good. A few questions and nits plus yetus/revive things to address.
Kicking off the tests in parallel with that.

milan-zededa commented on pkg/pillar/cmd/zedagent/handlelocation.go:
```diff
@@ -148,15 +152,13 @@ func publishLocationToController(locInfo *info.ZInfoLocation, iteration int) {
+		forcePeriodic, info.ZInfoTypes_ZiLocation)
 }
 
 func publishLocationToLocalServer(ctx *getconfigContext, locInfo *info.ZInfoLocation) {
```
Since you are now also refactoring a bit, would you rename this to `publishLocationToLPS`? Also any other occurrence of LocalServer can be renamed to LPS to avoid any confusion.

rouming replied:
@milan-zededa Sure? 'LocalServer' is used in many places all over the code. I can, but won't this only increase confusion if I change only the function name?

milan-zededa replied:
I wouldn't mind if you renamed all of it :)

rouming replied:
Not in this PR. Here I offer only one function rename! :)

milan-zededa commented on pkg/pillar/cmd/zedagent/handleconfig.go:
```diff
@@ -643,27 +647,66 @@ func getLatestConfig(getconfigCtx *getconfigContext, url string,
+		url = zedcloud.URLPathString(locURL, zedcloudCtx.V2API, devUUID, "config")
+
+		// If LOC configuration is outdated, then we get obsoleteConfig return
+		// value (see parseConfig() for details) and w
```
Looks like unfinished comment.

rouming replied:
Thanks. Supposed to be without 'and w'.

milan-zededa commented:
It seems that `dev_local_info.txt` is failing, which is an eden test related to LPS and could have been broken by this PR, hence worth investigating.

rouming commented:
Difference to previous version:
* A couple of fixes for the LOC case
* Yetus fixes, comments tweaks, code tweaks
* Added global rename of LocalProfileServer,LocalServer,LocalProfile,etc,etc to lps or LPS. @milan-zededa last 7 commits worth to review, hope this was the intention. The changes are rather mechanical with some obvious reverts in protocol/API files, so some of the LocalProfileServer are still there in the code

milan-zededa reviewed (approved):
LGTM, but I would suggest to squash this PR into a smaller number of commits for easier to navigate git history.

eriknordmark commented:
Looks like yetus is tripping on:
yetus: pkg/pillar/cmd/zedagent/handlelps.go#L118golangcilint: getconfigCtx.localServerMap undefined (type *getconfigContext has no field or method localServerMap) (typecheck)

eriknordmark commented on pkg/pillar/cmd/zedagent/handlelps.go:
```diff
@@ -0,0 +1,288 @@
+)
+
+const (
+	defaultLPSPort = "8888"
```
FWIW I suspect it is the renaming of the constants and other names which makes GH not realize this is a rename of a file; it reports it as a delete of the old name and a create of this file.
Don't see a way to make it treat it as a rename + set of content changes.

eriknordmark reviewed (approved):
Run eden again

eriknordmark reviewed (changes requested):
Looks like there are still comment comment suggestions plus yetus failures of the form
golangcilint: getconfigCtx.localServerMap undefined (type *getconfigContext has no field or method localServerMap) (typecheck)
to address then we can merge this.

rouming commented:
Difference to the previous version:
* Rebased on master
* Removed all commits related to the renaming of the LocalServer to LPS etc. I do not feel comfortable to commit these changes in this PR.

eriknordmark reviewed (approved):
LGTM

eriknordmark commented:
I'll kick off another run, but @rouming can you figure out why TestEdenScripts/port_forward fails 2/2 times in https://github.com/lf-edge/eve/actions/runs/4376395843/jobs/7658444261 and the eden tests run out of time?

rouming commented:
> I'll kick off another run, but @rouming can you figure out why TestEdenScripts/port_forward fails 2/2 times in https://github.com/lf-edge/eve/actions/runs/4376395843/jobs/7658444261 and the eden tests run out of time?

Can try

rouming commented:
> > I'll kick off another run, but @rouming can you figure out why TestEdenScripts/port_forward fails 2/2 times in https://github.com/lf-edge/eve/actions/runs/4376395843/jobs/7658444261 and the eden tests run out of time?
> 
> Can try

@eriknordmark Port forwarding does not work. As far as I can tell all nat iptables rules exist (2223->22), but indeed you can't ssh to the app using eve ip addr: $ ssh 192.168.0.10 -p2223, but directly app-vm accepts the connection: $ ssh 10.11.12.2, I will try to debug further.

These eden tests successfully passed on this PR on my local machine:
./eden test tests/eclient -e profile -v debug
./eden test tests/eclient -e radio_silence -v debug
./eden test tests/eclient -e dev_local_info -v debug

so I'm about to rebase and merge this PR.
